packages/valory/skills/proposal_collector_abci/behaviours.py

```diff
@@ -240,93 +250,69 @@ def _get_active_proposals(self) -> Generator[None, None, str]:
-    """
-    SelectProposal
-
-    Select the proposal on which to vote.
-    """
-
-    matching_round: Type[AbstractRound] = SelectProposalRound
-
-    def async_act(self) -> Generator:
-        """Do the act, supporting asynchronous execution."""
+        # Remove proposals for which the vote end block is in the past
+        current_block = yield from self.get_current_block()
+        if current_block is None:
+            return CollectActiveProposalsRound.BLOCK_RETRIEVAL_ERROR
 
-        with self.context.benchmark_tool.measure(self.behaviour_id).local():
-
-            # TODO: we enter this round after a successful tx_submission
-            # We need to check if that is the case and remove the first proposal
-            # from the list.
-            # TODO: we want to make sure we only vote towards the end of the voting period.
-            # But it would be great if we could get the voting intention right away, so we can
-            # display this on the frontend.
-
-            # Get all the governors from the current delegations
-            delegation_governors = [
-                d["governor_address"] for d in self.synchronized_data.delegations
-            ]
-
-            # Get the proposals where we can vote:
-            # - Active
-            # - Not voted before
-            # - Governor in the delegation list
-            votable_active_proposals = {
-                p_id: p
-                for p_id, p in self.synchronized_data.proposals.items()
-                if p["active"]
-                and not p["vote"]
-                and p["governor"]["id"].split(":")[-1] in delegation_governors
-            }
-
-            # Some proposals have an ETA, some dont
-            # We will prioritise those with it
-            active_proposals_with_eta = list(
-                filter(lambda p: p["eta"] != "", votable_active_proposals.values())
-            )
-            active_proposals_no_eta = list(
-                filter(lambda p: p["eta"] == "", votable_active_proposals.values())
-            )
-
-            # Sort by ETA
-            active_proposals_with_eta_sorted = sorted(
-                active_proposals_with_eta, key=lambda p: int(p["eta"]), reverse=False
-            )
-
-            # Sort by ID
-            active_proposals_no_eta_sorted = sorted(
-                active_proposals_no_eta, key=lambda p: int(p["id"]), reverse=False
-            )
-
-            # Add all sorted proposals
-            sorted_proposals = (
-                active_proposals_with_eta_sorted + active_proposals_no_eta_sorted
-            )
-
-            # Select the first proposal
-            proposal_id = sorted_proposals[0]["id"] if sorted_proposals else None
-
-            # Check whether we have delegations
-            if not proposal_id:
-                proposal_id = SelectProposalRound.NO_PROPOSAL
+        active_proposals = filter(
+            lambda ap: ap["end"]["number"] > current_block, active_proposals
+        )
```

Comment:
Why not also filter the proposals with the current block number and discard old ones? Isn't this dictionary going to keep growing infinitely if we do not discard the old proposals?

Reply:
This was on my list on issues as well, but I first need to sync up with Oak wrt frontend requirements. If we just remove past proposals that we voted on, the frontend wont be able to render them. One possible solution is to keep old proposals that we voted on for a number of blocks (i.e. ~a week).